File: C_plus/993_CousinsinBinaryTree.cpp
/*
 * @Author: yuan
 * @Date: 2021-05-05 13:52:52
 * @LastEditTime: 2021-05-05 13:52:52
 * @FilePath: /C_plus/993_CousinsinBinaryTree.cpp
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    void dfs(TreeNode *root, map<int, pair<TreeNode *, int> > &m, TreeNode *parent, int depth)
    {
        if (!root)
            return;
        m[root->val] = make_pair(parent, depth);
        dfs(root->left, m, root, depth + 1);
        dfs(root->right, m, root, depth + 1);
    }
    bool isCousins(TreeNode *root, int x, int y)
    {
        map<int, pair<TreeNode *, int> > m;
        dfs(root, m, nullptr, 0);
        return (m[x].first != m[y].first) && (m[x].second == m[y].second);
    }
};

File: Python/993_CousinsinBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        isX = False
        isY = False
        q = []
        q.append(root)
        while q:
            size = len(q)
            for i in range(size):
                p = q[0]
                q.pop(0)
                if p.val == x:
                    isX = True
                if p.val == y:
                    isY = True
                if p.left and p.right:
                    left = p.left.val
                    right = p.right.val
                    if (left == x and right == y) or (left == y and right == x):
                        return False
                if p.left:
                    q.append(p.left)
                if p.right:
                    q.append(p.right)
            if isX & isY:
                return True
            if isX | isY:
                return False

        return False
